WIP: Introduce ssl-impl module with psql ssl support

mxm opened the pull request:
- introduce new ssl-impl module
- introduce netty ssl handler
- expose configuration of handler in settings

Still missing unit and integration tests for the SSL communication (only manually tested so far).

mxm commented:
Just saw that tests are failing because they use the enterprise version. I think we need to either have a default SSL configuration for the tests which use the enterprise version or disable SSL for the tests which are not SSL-specific.

mfussenegger commented on ssl/src/main/java/io/crate/protocols/postgres/ssl/SslReqHandlerSupplier.java:
```diff
@@ -22,16 +22,31 @@
 
-    public SslHandlerProvider(Settings settings) {
-        this.sslReqHandler = new SslReqRejectingHandler(settings);
+    public SslReqHandlerSupplier(Settings settings) {
```
Looks like this could be a single factory method to create the SSLReqHandler instead of creating a instance of the Supplier itself?

mxm replied:
Indeed. Can be simplified.

mfussenegger commented on ssl/src/main/java/io/crate/protocols/postgres/ssl/SslReqHandlerUtils.java:
```diff
@@ -25,7 +25,7 @@
-public class SslHandlerUtils {
+public class SslReqHandlerUtils {
 
     public static SslReqHandler getDefault() {
```
Is this now only used in tests? If so I think we should remove it here and move it into the test or just inline the constructor call

mxm replied:
This is shared between the different handler implementations.

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslConfigurationException.java:
```diff
@@ -0,0 +1,33 @@
+        super(s);
+    }
+
+    SslConfigurationException(String s, Exception e) {
```
Would be a bit more verbose with the argument names - e.g. msg / cause

mxm replied:
+1

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslConfiguration.java:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    private enum StoreType {
+        trust,
```
please use all upper case for enum values

matriv replied:
I did that on purpose because I use them to construct exception messages, and since this enum is only for that I didn't bother to do something like TRUST("trust"), KEY("key")...

mfussenegger replied:
Maybe add a short comment about that? Seems a bit hacky

mxm replied:
You can still use something like `enumValue.name().toLowerCase();` in the message.

mxm replied:
Or add a method getName() to the enum which does that.

mxm replied:
or even overload toString ;)

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslConfiguration.java:
```diff
@@ -0,0 +1,248 @@
+    @VisibleForTesting
+    static class KeyStoreSettings {
+
+        KeyStore keyStore;
```
please mark attributes as final if possible

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslConfiguration.java:
```diff
@@ -0,0 +1,248 @@
+    @VisibleForTesting
+    static class TrustStoreSettings {
+
+        KeyStore trustStore;
```
please mark attributes as final if possible

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslCertificateHelper.java:
```diff
@@ -0,0 +1,124 @@
+import java.security.cert.X509Certificate;
+import java.util.*;
+
+public class SslCertificateHelper {
```
Does splitting this into SslCertificateHelper & SslConfiguration  really make sense?

mxm replied:
Seems like the two should be consolidated.

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslCertificateHelper.java:
```diff
@@ -0,0 +1,124 @@
+        final List<X509Certificate> trustedCerts = new ArrayList<X509Certificate>();
+
+        if(LOGGER.isDebugEnabled()) {
+            LOGGER.debug("No alias given, will trust all of the certificates in the store");
```
this log message seems odd as there is no check in regards to the aliases at all?

mxm replied:
Yes, particularly because if there are no aliases, then we will return an empty array of trusted certificates; this should lead to trusting _no_ certificate.

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslCertificateHelper.java:
```diff
@@ -0,0 +1,124 @@
+    private static boolean stripRootFromChain = true; //TODO check
+
+    public static X509Certificate[] exportRootCertificates(final KeyStore ks) throws KeyStoreException {
+        final List<String> aliases = toList(ks.aliases());
```
`toList` seems unnecessary as there is only a single loop done later on?

mxm replied:
Yes, this is not necessary.

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslCertificateHelper.java:
```diff
@@ -0,0 +1,124 @@
+        final List<String> aliases = toList(ks.aliases());
+        if (aliases.isEmpty()) {
+            String msg = "Keystore does not contain any aliases";
+            LOGGER.error(msg);
```
Think we don't need to log if the error is also propagated

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslCertificateHelper.java:
```diff
@@ -0,0 +1,124 @@
+        return trustedCerts.toArray(new X509Certificate[0]);
+    }
+
+    private static List<String> toList(final Enumeration<String> enumeration) {
```
can be replaced by using `Collections.list`

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslConfiguration.java:
```diff
@@ -0,0 +1,248 @@
+
+    private static SslContext buildSSLContext(final SslContextBuilder sslContextBuilder) throws SSLException {
+
+        final SecurityManager sm = System.getSecurityManager();
```
Do we need this? Thought we don't use the SecurityManager anyway

mxm replied:
It's not needed here.

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslConfiguration.java:
```diff
@@ -0,0 +1,248 @@
+            TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        trustFactory.init(trustStore);
+
+        // get the trust managers from the factory
```
Instead of these kind of obvious comments I think some more high-level explanation of what's going on would be more helpful  =)

Some method and class level docs would be great

mxm replied:
Will update the documentation and remove these self-explanatory comments :)

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslConfiguration.java:
```diff
@@ -0,0 +1,248 @@
+
+        // load the store
+        final KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+        keyStore.load(new FileInputStream(new File(keyStorePath)),
```
doesn't this leak the file descriptor?

mxm replied:
Not sure I understand. `load` is a void method.

mfussenegger replied:
The FileInputStream isn't closed, so the file descriptor isn't freed (immediately)

It eventually will be freed if finalize is called on the instance, but it would probably be cleaner to wrap it into a try block

mxm replied:
Thanks for the explanation. Pushing a fix.

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslConfiguration.java:
```diff
@@ -0,0 +1,248 @@
+
+        // initialize a key manager factory with the key store
+        KeyManagerFactory keyFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+        keyFactory.init(
```
this looks all very similar to the trustStore code - could we extract some common methods?

mxm replied:
Looks like the `KeyManagerFactory` and the `TrustManagerFactory` don't share a common interface.

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslReqConfiguringHandler.java:
```diff
@@ -0,0 +1,75 @@
+     * Constructs the Netty SslHandler which should be added as the first element of the pipeline.
+     */
+    SslHandler buildSSLHandler(ChannelPipeline pipeline) throws SSLException, CertificateException {
+        return SslConfiguration.buildSslContext(settings).newHandler(pipeline.channel().alloc());
```
Think we could consider injecting a `Supplier<SslContext>` in the constructor. Would decouple this class from the concrete ssl-context-building logic and also from the settings class.

Would also remove the need for a separate subclass for testing with the SelfSignedCertificate

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslConfiguration.java:
```diff
@@ -249,7 +249,12 @@ static KeyStoreSettings loadKeyStore(Settings settings) throws KeyStoreException
+
+        @Override
+        public String toString() {
+            return this.name().toLowerCase();
```
please always set an explicit locale to lower/uppercase calls. Otherwise with some locales this can lead to a strange behaviour, depending on how these names are used

mxm commented:
I've addressed your comments, apart from:

- unifying `SSLConfiguration` and `SslCertificateHelper` which still needs some work

mxm commented:
@mfussenegger and @matriv 

I've updated the PR. Here's a quick summary:

- add new certificates with small certificate chain
- add end-to-end integration test
- truststore is now optional
- add ssl enable flag 
- refactored `SslConfiguration` and removed `SslCertificateHelpers`.
- new tests for all used functionality
- bug fixes

mfussenegger commented on ssl-impl/src/main/java/io/crate/protocols/postgres/ssl/SslReqConfiguringHandler.java:
```diff
@@ -0,0 +1,88 @@
+        this(settings, new DefaultSslContextSupplier(settings));
+    }
+
+    public SslReqConfiguringHandler(Settings settings, Supplier<SslContext> sslContextSupplier) {
```
I think it would now be possible to drop the `Supplier` and require the SslContext directly since it's always created immediately

mxm replied:
Yes, a supplier is overhead. Removed.

mxm commented:
Thanks for the review! I've squashed all the individual commits. I will let the tests run before I merge.

>I think generally it would be good to prefer assertThat over assertTrue for improved assertion messages in case of failures.

I agree. Removed all `assertTrue`s.
